File: Guías Completas/Estructuras Algebraicas/Contracts/Python/Estructuras_Algebraicas.py
# Operaciones básicas en los números naturales

# Suma
def suma(a, b):
  return a + b

# Elemento neutro para la suma
elemento_neutro_suma = 0

# Resta
def resta(a, b):
  return a - b

# Elemento neutro para la resta
elemento_neutro_resta = 0

# Multiplicación
def multiplicacion(a, b):
  return a * b

# Elemento neutro para la multiplicación
elemento_neutro_multiplicacion = 1

# División
def division(a, b):
  return a / b

# Comprobación de la propiedad asociativa de la suma
def asociativa_suma(a, b, c):
  return (a + b) + c == a + (b + c)

# Comprobación de la propiedad conmutativa de la suma
def conmutativa_suma(a, b):
  return a + b == b + a

# Comprobación de la propiedad distributiva de la multiplicación sobre la suma
def distributiva_multiplicacion_sobre_suma(a, b, c):
  return a * (b + c) == a * b + a * c

# Ejemplo de uso:
resultado_suma = suma(2, 7)
print(resultado_suma) # Imprime 9

resultado_resta = resta(4, 6)
print(resultado_resta) # Imprime -2

resultado_multiplicacion = multiplicacion(7, 1)
print(resultado_multiplicacion) # Imprime 7

resultado_division = division(7, 2)
print(resultado_division) # Imprime 3.5

es_asociativa_suma = asociativa_suma(2, 7, 3)
print(es_asociativa_suma) # Imprime True

es_conmutativa_suma = conmutativa_suma(2, 7)
print(es_conmutativa_suma) # Imprime False

es_distributiva_multiplicacion_sobre_suma = distributiva_multiplicacion_sobre_suma(2, 3, 4)
print(es_distributiva_multiplicacion_sobre_suma) # Imprime True

File: Guías Completas/Estructuras Algebraicas/Contracts/Rust/Estructuras_Algebraicas.rs
// Suma
fn suma(a: i32, b: i32) -> i32 {
  return a + b;
}

// Resta
fn resta(a: i32, b: i32) -> i32 {
  return a - b;
}

// Multiplicación
fn multiplicacion(a: i32, b: i32) -> i32 {
  return a * b;
}

// División
fn division(a: i32, b: i32) -> f32 {
  return (a as f32) / (b as f32);
}

// Comprobación de la propiedad asociativa de la suma
fn asociativa_suma(a: i32, b: i32, c: i32) -> bool {
  return (a + b) + c == a + (b + c);
}

// Comprobación de la propiedad conmutativa de la suma
fn conmutativa_suma(a: i32, b: i32) -> bool {
  return a + b == b + a;
}

// Comprobación de la propiedad distributiva de la multiplicación sobre la suma
fn distributiva_multiplicacion_sobre_suma(a: i32, b: i32, c: i32) -> bool {
  return a * (b + c) == a * b + a * c;
}

fn main() {
  let resultado_suma = suma(2, 7);
  println!("{}", resultado_suma); // Imprime 9

  let resultado_resta = resta(4, 6);
  println!("{}", resultado_resta); // Imprime -2

  let resultado_multiplicacion = multiplicacion(7, 1);
  println!("{}", resultado_multiplicacion); // Imprime 7

  let resultado_division = division(7, 2);
  println!("{}", resultado_division); // Imprime 3.5

  let es_asociativa_suma = asociativa_suma(2, 7, 3);
  println!("{}", es_asociativa_suma); // Imprime true

  let es_conmutativa_suma = conmutativa_suma(2, 7);
  println!("{}", es_conmutativa_suma); // Imprime false

  let es_distributiva_multiplicacion_sobre_suma = distributiva_multiplicacion_sobre_suma(2, 3, 4);
  println!("{}", es_distributiva_multiplicacion_sobre_suma); // Imprime true
}
